heudiconv/parser.py

```diff
@@ -60,58 +62,87 @@ def find_files(regex, topdir=op.curdir, exclude=None,
             yield path
 
 
-def get_extracted_dicoms(fl):
-    """Given a list of files, possibly extract some from tarballs.
+def get_extracted_dicoms(
+        fl: typing.Iterable[str]
+        ) -> typing.ItemsView[typing.Optional[int], typing.List[str]]:
+    """Given a collection of files and/or directories, possibly extract 
+    some from tarballs.
 
-    For 'classical' heudiconv, if multiple tarballs are provided, they correspond
-    to different sessions, so here we would group into sessions and return
-    pairs  `sessionid`, `files`  with `sessionid` being None if no "sessions"
-    detected for that file or there was just a single tarball in the list
-    """
-    # TODO: bring check back?
-    # if any(not tarfile.is_tarfile(i) for i in fl):
-    #     raise ValueError("some but not all input files are tar files")
+    Parameters
+    ----------
+    fl : Iterable[str]
+        An iterable (e.g., list or tuple) of files to process.
+
+    Returns
+    -------
+    ItemsView[int | None, list[str]]
+        A tuple of keys (either integer or None) and list of strs representing
+          the absolute paths of files. 
 
-    # tarfiles already know what they contain, and often the filenames
-    # are unique, or at least in a unique subdir per session
-    # strategy: extract everything in a temp dir and assemble a list
-    # of all files in all tarballs
+    Notes
+    -----
+    For 'classical' heudiconv, if multiple archives are provided, they 
+    correspond to different sessions, so here we would group into sessions 
+    and return pairs `sessionid`, `files`  with `sessionid` being None if no 
+    "sessions" detected for that file or there was just a single tarball in the
+    list.
 
-    # cannot use TempDirs since will trigger cleanup with __del__
-    tmpdir = tempdirs('heudiconvDCM')
+    When contents of fl appear to be an unpackable archive, the contents are
+    extracted into utils.TempDirs(f'heudiconvDCM') and the mode of all 
+    extracted files is set to 700.
 
-    sessions = defaultdict(list)
-    session = 0
-    if not isinstance(fl, (list, tuple)):
+    When contents of fl are a list of files, they are treated as a single
+    session.
+    """
+    input_list_of_unpacked = any(fl) and all(
+        not t.endswith(_UNPACK_FORMATS) for t in fl
+        )
+    
+    if not (
+        input_list_of_unpacked or
+        all(t.endswith(_UNPACK_FORMATS) for t in fl)
+        ):
+        raise ValueError("Some but not all input files are archives.")
```

Comment:
I think prior code was handling that just fine -- and add non-archive to `None` session.  Is there a reason to introduce this breaking behavior?  I think it might have really bad manifestations (didn't check) e.g. by pointing to a folder with some tarballs and some files to just be ignored which we ignore if not dicom later (?) in the code, etc.